Add debug information for inline function parameters

nikita-nazarov opened the pull request:
Related issue: https://youtrack.jetbrains.com/issue/KT-55440

Previously during debugging Kotlin/Native applications you couldn't inspect the inline function or inline lambda parameters. This happened because the debug information for them wasn't generated at all. To fix this issue we have to store the information about which expressions were inlined instead of which parameters. This commit adds a new mapping, which allows storing the required information during the inlining phase. The mapping is then reused during the bitcode generation phase to add the debug information for inline function parameters.

This pr only handles the inline function parameters that were put on a stack. If a constant value is passed as a parameter, it still won't be shown. Although it is possible to add the debug information for such parameters, I decided to focus this pr only on stack-based ones.

I also have to mention that this issue also happens with extension receivers:
```
inline fun Int.foo() {
   println() // Breakpoint here, this won't be visible
}
```
But this topic is also out of scope of the pull request.

@madsager

nikita-nazarov commented on kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:
```diff
@@ -2102,6 +2103,22 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
+        val inlineFunctionSymbol = block.inlineFunctionSymbol ?: return
+        val function = inlineFunctionSymbol.owner
+        for (parameter in function.valueParameters) {
+            if (parameter.type.isFunctionTypeOrSubtype()) continue
```
We don't add any information for inline lambdas since they don't have any address. However we may add some information for noinline limbdas, but would it be of any value during debugging?

kunyavskiy replied:
I think it's fine for now. It probably can be useful if there is a good way to show it, but anyway, you can't see anything meaningful from function reference for now.

kunyavskiy commented on kotlin-native/backend.native/tests/debugger/src/test/kotlin/org/jetbrains/kotlin/native/test/debugger/LldbTests.kt:
```diff
@@ -495,4 +495,44 @@ class LldbTests {
+        fun main() {
+            val tmp1 = 1
+            val tmp2 = 2
+            foo(tmp1, tmp2) { pix, piy, piz ->
```
It seams, that the current code works only if the variable is passed to an argument. 

For example, if I change the code to `foo(1, 2)` variables disappear. Also p2 disappear if use `foo(tmp1, tmp1+tmp1)`. Can you please check it, and add more complex tests?

kunyavskiy commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/inline/FunctionInlining.kt:
```diff
@@ -227,6 +229,13 @@ class FunctionInlining(
+        private fun saveParameterToExpressionMapping(callee: IrFunction, copiedCallee: IrFunction) {
+            for ((parameter, copiedParameter) in callee.valueParameters.zip(copiedCallee.valueParameters)) {
+                val expression = substituteMap[copiedParameter] ?: continue
+                context.mapping.inlinedParameterToExpression[parameter] = expression
```
Did I understand correctly that you are using the inline function parameter as the key in the map? How it's supposed to work if the function is called several times? Can you please add a test on this case, if I'm missing something and it works?

kunyavskiy commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/inline/FunctionInlining.kt:
```diff
@@ -199,6 +199,8 @@ class FunctionInlining(
             newStatements.addAll(evaluationStatements)
             statements.mapTo(newStatements) { it.transform(transformer, data = null) as IrStatement }
 
+            saveParameterToExpressionMapping(callee, copiedCallee)
```
There should be variables for these parameters in most cases. I think they are not shown now, as they are temporary variables. What do you think about creating a separate origin for them? I'm not sure about the name and scope, but this should be an easier solution, at least for cases, where variables are not optimized-out.

nikita-nazarov commented:
So I've added a solution that follows the new approach. Now, when the debug flag is enabled, Kotlin Native compiler always generates temporary variables for all arguments of an inline function. These variables will have the same name as their corresponding argument, and the debug information will be added for them automatically.

I've added the 'revert' commit to this branch just to keep the code of the previous approach in case we will need it again.

nikita-nazarov commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/inline/FunctionInlining.kt:
```diff
@@ -602,43 +603,61 @@ class FunctionInlining(
+                }
+            )
+
+            if (alwaysCreateTemporaryVariablesForArguments) {
```
I'm not sure if we should name the temporary variables like arguments only if the `alwaysCreateTemporaryVariablesForArguments` option is enabled or in all cases.

kunyavskiy commented on kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanLoweringPhases.kt:
```diff
@@ -186,7 +186,12 @@ internal val wrapInlineDeclarationsWithReifiedTypeParametersLowering = makeKonan
+            FunctionInlining(
+                    context,
+                    NativeInlineFunctionResolver(context, generationState),
+                    alwaysCreateTemporaryVariablesForArguments = generationState.config.debug
```
please use `generationState.shouldContainDebugInfo` instead

kunyavskiy commented on compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/inline/FunctionInlining.kt:
```diff
@@ -602,43 +603,61 @@ class FunctionInlining(
+                    return@forEach
+                }
 
                 val argumentExtracted = !argument.argumentExpression.isPure(false, context = context)
```
For now, in fact we have to cases 
1. Directly substitute
2. Create temporary variable. 

For now logic on when which is happening is distributed all over the function code. It would be nice to collect it in one place. Like add a variable "shouldInlineArgumentValue", and just check over it and the end, without returns in the middle of function.
